src/pybind/mgr/zabbix/module.py

```diff
@@ -86,14 +86,23 @@ def __init__(self, *args, **kwargs):
         self.event = Event()
 
     def init_module_config(self):
+        firste = None
         for key, default in self.config_keys.items():
-            value = self.get_localized_config(key, default)
-            if value is None:
-                raise RuntimeError('Configuration key {0} not set; "ceph '
+            try:
+                value = self.get_localized_config(key, default)
+                if value is not None:
+                    self.set_config_option(key, value)
```

Comment:
So this is now the same logic as the existing code, except that it is storing up the first exception and finishing the loop before re-raising it.

I don't see how this is fixing anything?

Reply:
Oh, i found that in my development environment using the vstart to start mgr when only set zabbix_host and identifier. The keypoint is that when in the fool loop when first module init,the first key will be zabbix_host,once it is not be set,the init module will be raising exeption while bypath the later default set!
Then later i set the host and identifer, the other default key indeed is not set .so In the send func(trigger by the command zabbix send),it will access nil index,lead to fail finally.


The process is 
` /home/ansibler/ceph/build/bin/ceph -c /home/ansibler/ceph/build/ceph.conf mgr module enable zabbix
 /home/ansibler/ceph/build/bin/ceph -c /home/ansibler/ceph/build/ceph.conf zabbix config-show
/home/ansibler/ceph/build/bin/ceph -c /home/ansibler/ceph/build/ceph.conf mgr module l
/home/ansibler/ceph/build/bin/ceph zabbix config-set zabbix_host 192.168.122.36
/home/ansibler/ceph/build/bin/ceph zabbix config-set identifier hostname(kernel)
/home/ansibler/ceph/build/bin/ceph zabbix send 
`
The result is failed.
Log is 
`1211:2017-11-24 17:55:32.350 7f55fb5da700  1 mgr init Loading python module 'zabbix'
1264:2017-11-24 17:55:32.780 7f55df346700  1 mgr load Constructed class from module: zabbix
1267:2017-11-24 17:55:32.781 7f55dbb3f700 -1 log_channel(cluster) log [ERR] : Unhandled exception from module 'zabbix' while running on mgr.x: Configuration key zabbix_host not set; "ceph config-key set mgr/zabbix/zabbix_host <value>"
1268:2017-11-24 17:55:32.781 7f55dbb3f700 -1 zabbix.serve:
1271:  File "/home/ansibler/ceph/src/pybind/mgr/zabbix/module.py", line 257, in serve
1273:  File "/home/ansibler/ceph/src/pybind/mgr/zabbix/module.py", line 95, in init_module_config
1275:RuntimeError: Configuration key zabbix_host not set; "ceph config-key set mgr/zabbix/zabbix_host <value>"
1382:2017-11-24 17:55:40.065 7f55deb45700  1 -- 192.168.122.2:6800/884 <== client.4182 192.168.122.2:0/4079037767 1 ==== command(tid 0: {"prefix": "zabbix config-show", "target": ["mgr", ""]}) v1 ==== 79+0+0 (3207660371 0 0) 0x55cb225f7a00 con 0x55cb225a8600
1383:2017-11-24 17:55:40.065 7f55deb45700  0 log_channel(audit) log [DBG] : from='client.4182 192.168.122.2:0/4079037767' entity='client.admin' cmd=[{"prefix": "zabbix config-show", "target": ["mgr", ""]}]: dispatch
1403:2017-11-24 17:55:40.065 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-set'
1404:2017-11-24 17:55:40.065 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-show'
1667:2017-11-24 17:56:33.262 7f55deb45700  1 -- 192.168.122.2:6800/884 <== client.4171 192.168.122.2:0/38229306 1 ==== command(tid 0: {"prefix": "zabbix config-set", "key": "zabbix_host", "value": "192.168.122.36", "target": ["mgr", ""]}) v1 ==== 127+0+0 (1499170302 0 0) 0x55cb221fcc00 con 0x55cb22665800
1668:2017-11-24 17:56:33.263 7f55deb45700  0 log_channel(audit) log [DBG] : from='client.4171 192.168.122.2:0/38229306' entity='client.admin' cmd=[{"prefix": "zabbix config-set", "key": "zabbix_host", "value": "192.168.122.36", "target": ["mgr", ""]}]: dispatch
1688:2017-11-24 17:56:33.263 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-set'
1689:2017-11-24 17:56:33.264 7f55df346700  1 -- 192.168.122.2:0/2080213087 --> 192.168.122.2:40231/0 -- mon_command({"prefix":"config-key set","key":"mgr/zabbix/x/zabbix_host","val":"192.168.122.36"} v 0) v1 -- 0x55cb22557440 con 0
1691:2017-11-24 17:56:33.289 7f55fb5da700  1 -- 192.168.122.2:0/2080213087 <== mon.1 192.168.122.2:40231/0 71 ==== mon_command_ack([{"prefix":"config-key set","key":"mgr/zabbix/x/zabbix_host","val":"192.168.122.36"}]=0 set mgr/zabbix/x/zabbix_host v0) v1 ==== 145+0+0 (2511846422 0 0) 0x55cb22557440 con 0x55cb200f6c00
1773:2017-11-24 17:56:47.662 7f55deb45700  1 -- 192.168.122.2:6800/884 <== client.4191 192.168.122.2:0/1930969518 1 ==== command(tid 0: {"prefix": "zabbix config-set", "key": "identifier", "value": "kernel", "target": ["mgr", ""]}) v1 ==== 118+0+0 (846473440 0 0) 0x55cb221fc800 con 0x55cb22666a00
1774:2017-11-24 17:56:47.663 7f55deb45700  0 log_channel(audit) log [DBG] : from='client.4191 192.168.122.2:0/1930969518' entity='client.admin' cmd=[{"prefix": "zabbix config-set", "key": "identifier", "value": "kernel", "target": ["mgr", ""]}]: dispatch
1794:2017-11-24 17:56:47.663 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-set'
1795:2017-11-24 17:56:47.664 7f55df346700  1 -- 192.168.122.2:0/2080213087 --> 192.168.122.2:40231/0 -- mon_command({"prefix":"config-key set","key":"mgr/zabbix/x/identifier","val":"kernel"} v 0) v1 -- 0x55cb22556fc0 con 0
1797:2017-11-24 17:56:47.684 7f55fb5da700  1 -- 192.168.122.2:0/2080213087 <== mon.1 192.168.122.2:40231/0 86 ==== mon_command_ack([{"prefix":"config-key set","key":"mgr/zabbix/x/identifier","val":"kernel"}]=0 set mgr/zabbix/x/identifier v0) v1 ==== 135+0+0 (408748241 0 0) 0x55cb22557d40 con 0x55cb200f6c00
1911:2017-11-24 17:57:08.413 7f55deb45700  1 -- 192.168.122.2:6800/884 <== client.4163 192.168.122.2:0/1544083648 1 ==== command(tid 0: {"prefix": "zabbix send", "target": ["mgr", ""]}) v1 ==== 72+0+0 (2174970189 0 0) 0x55cb2260ca00 con 0x55cb226c9200
1912:2017-11-24 17:57:08.413 7f55deb45700  0 log_channel(audit) log [DBG] : from='client.4163 192.168.122.2:0/1544083648' entity='client.admin' cmd=[{"prefix": "zabbix send", "target": ["mgr", ""]}]: dispatch
1932:2017-11-24 17:57:08.413 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-set'
1933:2017-11-24 17:57:08.413 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-show'
1934:2017-11-24 17:57:08.413 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix send'
1935:2017-11-24 17:57:08.416 7f55df346700  0 mgr[zabbix] Exception when sending: 'zabbix_sender'
1211:2017-11-24 17:55:32.350 7f55fb5da700  1 mgr init Loading python module 'zabbix'
1264:2017-11-24 17:55:32.780 7f55df346700  1 mgr load Constructed class from module: zabbix
1267:2017-11-24 17:55:32.781 7f55dbb3f700 -1 log_channel(cluster) log [ERR] : Unhandled exception from module 'zabbix' while running on mgr.x: Configuration key zabbix_host not set; "ceph config-key set mgr/zabbix/zabbix_host <value>"
1268:2017-11-24 17:55:32.781 7f55dbb3f700 -1 zabbix.serve:
1271:  File "/home/ansibler/ceph/src/pybind/mgr/zabbix/module.py", line 257, in serve
1273:  File "/home/ansibler/ceph/src/pybind/mgr/zabbix/module.py", line 95, in init_module_config
1275:RuntimeError: Configuration key zabbix_host not set; "ceph config-key set mgr/zabbix/zabbix_host <value>"
1382:2017-11-24 17:55:40.065 7f55deb45700  1 -- 192.168.122.2:6800/884 <== client.4182 192.168.122.2:0/4079037767 1 ==== command(tid 0: {"prefix": "zabbix config-show", "target": ["mgr", ""]}) v1 ==== 79+0+0 (3207660371 0 0) 0x55cb225f7a00 con 0x55cb225a8600
1383:2017-11-24 17:55:40.065 7f55deb45700  0 log_channel(audit) log [DBG] : from='client.4182 192.168.122.2:0/4079037767' entity='client.admin' cmd=[{"prefix": "zabbix config-show", "target": ["mgr", ""]}]: dispatch
1403:2017-11-24 17:55:40.065 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-set'
1404:2017-11-24 17:55:40.065 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-show'
1667:2017-11-24 17:56:33.262 7f55deb45700  1 -- 192.168.122.2:6800/884 <== client.4171 192.168.122.2:0/38229306 1 ==== command(tid 0: {"prefix": "zabbix config-set", "key": "zabbix_host", "value": "192.168.122.36", "target": ["mgr", ""]}) v1 ==== 127+0+0 (1499170302 0 0) 0x55cb221fcc00 con 0x55cb22665800
1668:2017-11-24 17:56:33.263 7f55deb45700  0 log_channel(audit) log [DBG] : from='client.4171 192.168.122.2:0/38229306' entity='client.admin' cmd=[{"prefix": "zabbix config-set", "key": "zabbix_host", "value": "192.168.122.36", "target": ["mgr", ""]}]: dispatch
1688:2017-11-24 17:56:33.263 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-set'
1689:2017-11-24 17:56:33.264 7f55df346700  1 -- 192.168.122.2:0/2080213087 --> 192.168.122.2:40231/0 -- mon_command({"prefix":"config-key set","key":"mgr/zabbix/x/zabbix_host","val":"192.168.122.36"} v 0) v1 -- 0x55cb22557440 con 0
1691:2017-11-24 17:56:33.289 7f55fb5da700  1 -- 192.168.122.2:0/2080213087 <== mon.1 192.168.122.2:40231/0 71 ==== mon_command_ack([{"prefix":"config-key set","key":"mgr/zabbix/x/zabbix_host","val":"192.168.122.36"}]=0 set mgr/zabbix/x/zabbix_host v0) v1 ==== 145+0+0 (2511846422 0 0) 0x55cb22557440 con 0x55cb200f6c00
1773:2017-11-24 17:56:47.662 7f55deb45700  1 -- 192.168.122.2:6800/884 <== client.4191 192.168.122.2:0/1930969518 1 ==== command(tid 0: {"prefix": "zabbix config-set", "key": "identifier", "value": "kernel", "target": ["mgr", ""]}) v1 ==== 118+0+0 (846473440 0 0) 0x55cb221fc800 con 0x55cb22666a00
1774:2017-11-24 17:56:47.663 7f55deb45700  0 log_channel(audit) log [DBG] : from='client.4191 192.168.122.2:0/1930969518' entity='client.admin' cmd=[{"prefix": "zabbix config-set", "key": "identifier", "value": "kernel", "target": ["mgr", ""]}]: dispatch
1794:2017-11-24 17:56:47.663 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-set'
1795:2017-11-24 17:56:47.664 7f55df346700  1 -- 192.168.122.2:0/2080213087 --> 192.168.122.2:40231/0 -- mon_command({"prefix":"config-key set","key":"mgr/zabbix/x/identifier","val":"kernel"} v 0) v1 -- 0x55cb22556fc0 con 0
1797:2017-11-24 17:56:47.684 7f55fb5da700  1 -- 192.168.122.2:0/2080213087 <== mon.1 192.168.122.2:40231/0 86 ==== mon_command_ack([{"prefix":"config-key set","key":"mgr/zabbix/x/identifier","val":"kernel"}]=0 set mgr/zabbix/x/identifier v0) v1 ==== 135+0+0 (408748241 0 0) 0x55cb22557d40 con 0x55cb200f6c00
1911:2017-11-24 17:57:08.413 7f55deb45700  1 -- 192.168.122.2:6800/884 <== client.4163 192.168.122.2:0/1544083648 1 ==== command(tid 0: {"prefix": "zabbix send", "target": ["mgr", ""]}) v1 ==== 72+0+0 (2174970189 0 0) 0x55cb2260ca00 con 0x55cb226c9200
1912:2017-11-24 17:57:08.413 7f55deb45700  0 log_channel(audit) log [DBG] : from='client.4163 192.168.122.2:0/1544083648' entity='client.admin' cmd=[{"prefix": "zabbix send", "target": ["mgr", ""]}]: dispatch
1932:2017-11-24 17:57:08.413 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-set'
1933:2017-11-24 17:57:08.413 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix config-show'
1934:2017-11-24 17:57:08.413 7f55deb45700  1 mgr.server handle_command pyc_prefix: 'zabbix send'
1935:2017-11-24 17:57:08.416 7f55df346700  0 mgr[zabbix] Exception when sending: 'zabbix_sender'`